singularity-eos/eos/eos_eospac.hpp

```diff
@@ -72,103 +72,1156 @@ class EOSPAC : public EosBase<EOSPAC> {
+                         dTde.data(), "TofRE", Verbosity::Quiet);
+
+      table = PofRT_table_;
+      eosSafeInterpolate(&table, nelements, R.data(), T.data(), P.data(), dPdr, dPdT,
+                         "PofRT", Verbosity::Quiet);
+
+      for (int i = 0; i < nelements; ++i) {
+        int j = istart + i;
+        pressures[j] = pressureFromSesame(P[i]);
+      }
+    }
+  }
+
+  template <typename RealIndexer, typename ConstRealIndexer, typename LambdaIndexer>
+  inline void
+  PressureFromDensityInternalEnergy(ConstRealIndexer &&rhos, ConstRealIndexer &&sies,
+                                    RealIndexer &&pressures, Real *scratch, const int num,
+                                    LambdaIndexer &&lambdas) const {
+    using namespace EospacWrapper;
+    EOS_REAL *R = scratch;
+    EOS_REAL *E = scratch + num;
+    EOS_REAL *T = scratch + 2 * num;
+    EOS_REAL *dTdr = scratch + 3 * num;
+    EOS_REAL *dTde = scratch + 4 * num;
+    EOS_REAL *P = scratch + 5 * num;
+    EOS_REAL *dPdr = dTdr;
+    EOS_REAL *dPdT = dTde;
+
+    for (int i = 0; i < num; ++i) {
+      R[i] = rhos[i];
+      E[i] = sieToSesame(sies[i]);
+    }
+
+    EOS_INTEGER table = TofRE_table_;
+    eosSafeInterpolate(&table, num, R, E, T, dTdr, dTde, "TofRE", Verbosity::Quiet);
+
+    table = PofRT_table_;
+    eosSafeInterpolate(&table, num, R, T, P, dPdr, dPdT, "PofRT", Verbosity::Quiet);
+
+    for (int i = 0; i < num; ++i) {
+      pressures[i] = pressureFromSesame(P[i]);
+    }
+  }
+
+  template <int CHUNK_SIZE = 64, typename RealIndexer, typename ConstRealIndexer,
+            typename LambdaIndexer>
+  inline void SpecificHeatFromDensityTemperature(ConstRealIndexer &&rhos,
+                                                 ConstRealIndexer &&temperatures,
+                                                 RealIndexer &&cvs, const int num,
+                                                 LambdaIndexer &&lambdas) const {
+    using namespace EospacWrapper;
+    int nchunks = (num / CHUNK_SIZE) + (num % CHUNK_SIZE ? 1 : 0);
+    std::array<EOS_REAL, CHUNK_SIZE> R;
+    std::array<EOS_REAL, CHUNK_SIZE> T;
+    std::array<EOS_REAL, CHUNK_SIZE> E;
+    std::array<EOS_REAL, CHUNK_SIZE> DEDT;
+    std::array<EOS_REAL, CHUNK_SIZE> DEDR;
+
+    for (int ichunk = 0; ichunk < nchunks; ++ichunk) {
+      int istart = ichunk * CHUNK_SIZE;
+      int iend = std::min(num, istart + CHUNK_SIZE);
+      int nelements = iend - istart;
+
+      for (int i = 0; i < nelements; ++i) {
+        int j = istart + i;
+        R[i] = rhos[j];
+        T[i] = temperatureToSesame(temperatures[j]);
+      }
+
+      EOS_INTEGER table = EofRT_table_;
+      eosSafeInterpolate(&table, nelements, R.data(), T.data(), E.data(), DEDR.data(),
+                         DEDT.data(), "EofRT", Verbosity::Quiet);
+
+      for (int i = 0; i < nelements; ++i) {
+        int j = istart + i;
+        cvs[j] =
+            cvFromSesame(std::max(DEDT[i], 0.0)); // Here we do something to the data!
```

Comment:
I think we should get rid of this `max` but again since this was here before, I don't think it needs to be changed in this MR. Stuff like this should be handled in a production hardening layer I think.

Reply:
Again like you said, larger discussion. But one last comment and I'll leave it to another time.

My changes to the Gruneisen EOS were mainly targeted at sanitizing inputs to produce realistic outputs. Since there's no way to _a priori_ know what inputs will lead to bad inputs for sesame tables, I feel that it's not necessarily our place to modify what the EOS returns _without the user explicitly asking for it_. That's why I argue for a production hardening modifier to handle some of these things. The behavior of the modifier can be documented so that the user knows what they are guaranteed to be handed. I feel that without asking for a modifier like this, the user should be guaranteed the best representation of the EOS itself even if that data is garbage.

But again, this needs a larger discussion beyond the scope of this MR.